server/mqtt.go

```diff
@@ -3144,7 +3154,52 @@ func mqttPubTrace(pp *mqttPublish) string {
 //
 // Runs from the client's readLoop.
 // No lock held on entry.
-func (s *Server) mqttProcessPub(c *client, pp *mqttPublish) error {
+func (s *Server) mqttProcessPub(c *client, pp *mqttPublish, trace bool) error {
+	qos := mqttGetQoS(pp.flags)
+
```

Comment:
nit: you could capture `sess := c.mqtt.sess` to avoid repeating `c.mqtt.sess` (which is being used quite a bit in this function).